### PR TITLE
Add more parsers

### PR DESCRIPTION
**Pull request by All4Gis**
Implemented commented parser. TargetWidth fails because is this commented.

### CODE REVIEW

**Comment by paretech** [comment]
Thank you @All4Gis! PRs are most appreciated!

Checkout aebca07, I added 19 new tests to test_misb_0601.py. Without your pull request 19 fail, with 11 fail, with and TargetWidth uncommented 10 fail. Can you see if you can get the remaining 11 tests to pass or identify the cause to determine how to proceed?

**Comment by All4Gis** [comment]
Perfect! I'll do the tests for these new parsers

**Comment by paretech** [comment]
@All4Gis I have added basic tests for all tags of ST 0601 to test_misb_0601.py. 

I did skip/omitt a few, mostly those that did not include an example in Section 8 of MISB 0601.11 or those related to nested local sets. For instances I skipped, I did leave a comment, but those are silent from an automated test perspective.

**Comment by All4Gis** [comment]
Add a lot of new parsers.This parsers work using a video,but the videos that I have don't all have all metadata values.

The test result is : 
Ran 92 tests in 0.016s
FAILED (failures=10, errors=26)

Check this part if you can.

**Comment by paretech** [comment]
If I recall correctly, I think we can simply remove future.py from the master branch.

File: README.rst [hunk 1]
```diff
@@ -136,4 +136,8 @@ Documentation is available at https://paretech.github.io/klvdata.
 
 Contributing
 ------------
-Contributions are welcome!
\ No newline at end of file
+Contributions are welcome!
+
+Contributors List:
+------------
+- [Fran Raga](https://github.com/All4Gis)
```
Comment (paretech) [inline README.rst]: Absolutely!

File: klvdata/future.py [hunk 2]
```diff
@@ -67,7 +67,7 @@ class PlatformHeadingAngle(MappedConverterElement):
     # TODO: Design problem or thought needed to better structure converters
     #       to support validation during parsing, getting and setting.
     min_value, max_value, units = 0, 360, 'degrees'
-    min_length, max_length, signed = 2, 2, False
+    min_length, max_length, signed = 2, 2, True
```
Comment (paretech) [inline klvdata/future.py]: Is future.py still being used? I thought all the action was happening in misb0601.py and that the tests were setup to import from misb0601.py.

Comment (All4Gis) [reply klvdata/future.py]: Is better remove future.py in my opinion,but i don't want change completly the project structure,Remove this .py?

Comment (All4Gis) [reply klvdata/future.py]: And some conversions don`t work correctly, such as the checksum. I have to fix it

**Comment by All4Gis** [comment]
I can't get some tests to work!!
Ran 91 tests in 0.017s

FAILED (failures=20, errors=15)

**Comment by paretech** [comment]
Understood about the failing tests. I have a couple busy days ahead, I'll dig into this later this weekend or early next week. Thank you for your contributions! Have a great weekend.

**Comment by paretech** [comment]
Good afternoon @All4Gis, I'm looking into some of the reported test failures and errors you reported while working on this PR. I'll be using UASLSVersionNumber as an example but I think the following suggestions may help you get other element parser tests to pass.

If we look at http://www.gwg.nga.mil/misb/docs/standards/ST0601.11.pdf Tag 65 (Section 8.65) we will see that this item is a single byte mapping 0x00-0xFF to 1-255. So in our UASLSVersionNumber code, if we replace `StringElementParser` with `MappedElementParser` and add _domain = (0, 2**8-1) and _range = (0, 2**8-1) the associated test in test_misb_0601.py will pass. 

I think you can apply this same concept to GenericFlagData01, IcingDetected, LaserPRFCode, OperationalMode, SensorFieldOfViewName, TargetErrorEstimateLE90, TargetTrackGateHeight, TargetTrackGateWidth, TargetWidthExtended, WeaponFired, and WeaponLoad. Do note that I expect you may need to make some minor updates to the test cases along the way. 

Most of the other errors and exceptions are related to computational issues in the existing data structures and common.py. Some, like the one “int to big to convert” are issues that I am familiar with (see issue #4) and associated with a auto determined sign error. I plan to address these issues through a combination of architectural changes and expansion of unit level test cases.

Thank you again for your help, I hope this feedback helps you get 11 more cases passing or to the state of the other known issues. 

-Matt

**Comment by All4Gis** [comment]
if change this parsers using your suggestion, 
```
@UASLocalMetadataSet.add_parser
class UASLSVersionNumber(MappedElementParser):
    key = b'\x41'
    TAG = 64
    UDSKey = "-"
    LDSName = "UAS Datalink LS Version Number"
    ESDName = "ESD ICD Version"
    UDSName = ""
    _domain = (0, 28-1)
    _range = (0, 28-1)
    units = 'number'
```
the error is the ` OverflowError: int too big to convert`,then is better fix this problem #4  before change the test ,i think.

File: .gitignore [hunk 3]
```diff
@@ -110,3 +110,8 @@ ENV/
 
 # mypy
 .mypy_cache/
+.pydevproject
```
Comment (paretech) [inline .gitignore]: I'm unfamiliar with these enteries. Do they belong in a project level ignore or more in a local/global configuration?

File: klvdata/future.py [hunk 4]
```diff
@@ -1,333 +0,0 @@
-#!/usr/bin/env python3
```
Comment (paretech) [inline klvdata/future.py]: +1, nice to see this cleaned up (removed).

File: klvdata/misb0601.py [hunk 5]
```diff
@@ -59,6 +59,11 @@ class Checksum(BytesElementParser):
     Initialized from bytes value as BytesValue.
     """
     key = b'\x01'
+    TAG = 1
```
Comment (paretech) [inline klvdata/misb0601.py]: If possible, can we keep the scope of this PR down to adding just the additional element parsers and not additional functionality? I look forward to the day where klvdata can handle these features but I think it would be great if this PR focused on adding the missing parsers for which there is already support for (i.e. local set, using the existing Element Parsers and ST 0102 Nest Local Set).

File: klvdata/misb0601.py [hunk 6]
```diff
@@ -68,9 +73,14 @@ class PrecisionTimeStamp(DateTimeElementParser):
     Precision Timestamp represented in the number of microseconds elapsed
     since midnight (00:00:00), January 1, 1970 not including leap seconds.
 
-    See MISB ST 0603 for additional details.
+    See MISB ST 0601.11 for additional details.
```
Comment (paretech) [inline klvdata/misb0601.py]: The MISB standards are mostly active documents with revisions. Unless a specifc revision is critical, citing just the document number implies that the reader is intended to pull from the data management system (i.e. source) the latest revision. Also, if a revision is stated but isn't critical, the cost of maintaining this code/text increases as it will need to change everytime a revision to the cited document occurs. 

To summarize, unless critical, please cite standards using document numbers excluding revision identifiers.

File: test/test_misb_0601.py [hunk 7]
```diff
@@ -854,22 +855,22 @@ def test_PlatformAngleOfAttackFull(self):
...
+#         self.assertEqual(bytes(PlatformSideslipAngleFull(example_ls_packet[2:])), example_ls_packet)
 
     # Tag 94 (0x5E) MIIS Core Identifier (MISB ST 1204)
 
     # Tag 95 (0x5F) SAR Motion Imagery (MISB ST 1206) Local Set
 
     def test_TargetWidthExtended(self):
         # Example value and packet per MISB ST 0601.11, Section 8 "Conversions and Mappings of Metadata Types".
-        example_value = 13_898.5463
+        example_value = 13898.5463
```
Comment (paretech) [inline test/test_misb_0601.py]: The '_' was added for readability, it does not impact the value when executed.

File: klvdata/misb0601.py (hunk 6 above)
Comment (All4Gis) [reply klvdata/misb0601.py]: Ah okey!you know better those standards

File: klvdata/misb0601.py (hunk 5 above)
Comment (All4Gis) [reply klvdata/misb0601.py]: as you want. In this way the parsers are complete, I think

File: .gitignore (hunk 3 above)
Comment (All4Gis) [reply .gitignore]: Yes,is my local configuration,sorry!I'm use eclipse and pydev for develop.

File: test/test_misb_0601.py (hunk 7 above)
Comment (All4Gis) [reply test/test_misb_0601.py]: I thought it was a mistake, because the original ( standar ) value is 13898.5463

**Comment by All4Gis** [comment]
Commits ad6b114 and ef190af, fix tickets  #7 and #8

File: klvdata/misb0601.py [hunk 8]
```diff
@@ -851,7 +851,7 @@ class PlatformMagneticHeading(MappedElementParser):
 @UASLocalMetadataSet.add_parser
 class UASLSVersionNumber(StringElementParser):
     key = b'\x41'
-    TAG = 64
+    TAG = 65
```
Comment (paretech) [inline klvdata/misb0601.py]: +1

Comment (paretech) [reply klvdata/misb0601.py]: Is the change from StringElementParser to BytesElementParser related to the change in commit ef190afe2a237cc6f52a8a78c18888a0ab59df94? 

What happens if you use the MappedElementParser with _domain=(0, 2**8-1) and _range=(0, 255)?

**Comment by paretech** [comment]
@All4Gis, I merged PR #9 into master. It includes various corrections to MISB ST 0601 tests including SensorEastVelocity, GroundRange, OperationalMode, SlantRange, PlatformPitchAngle, IcingDetected, SensorNorthVelocity, TargetErrorEstimateCE90, TargetErrorEstimateLE90, and EventStartTime.

**Comment by All4Gis** [comment]
mm.. @paretech but any test not work using your changes:
 This is my complet report:

```
Finding files... done.
Importing test modules ... done.

======================================================================
ERROR: test_OutsideAirTemperature (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 420, in test_OutsideAirTemperature
    self.assertEqual(bytes(OutsideAirTemperature(example_value)), example_ls_packet)
TypeError: Can't instantiate abstract class OutsideAirTemperature with abstract methods _domain, _range

======================================================================
ERROR: test_TargetTrackGateWidth (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 456, in test_TargetTrackGateWidth
    self.assertEqual(bytes(TargetTrackGateWidth(example_value)), example_ls_packet)
TypeError: Can't instantiate abstract class TargetTrackGateWidth with abstract methods _domain, _range

======================================================================
FAIL: test_AlternatePlatformEllipsoidHeightExtended (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 925, in test_AlternatePlatformEllipsoidHeightExtended
    self.assertEqual(bytes(AlternatePlatformEllipsoidHeightExtended(example_value)), example_ls_packet)
AssertionError: b'i\x0823456.24' != b'i\x03/\x92\x1e'

======================================================================
FAIL: test_DensityAltitudeExtended (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 907, in test_DensityAltitudeExtended
    self.assertEqual(bytes(DensityAltitudeExtended(example_value)), example_ls_packet)
AssertionError: b'g\x0823456.24' != b'g\x03/\x92\x1e'

======================================================================
FAIL: test_GenericFlagData01 (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 492, in test_GenericFlagData01
    self.assertEqual(bytes(GenericFlagData01(example_value)), example_ls_packet)
AssertionError: b'/1\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00[152 chars]\x00' != b'/\x011'

======================================================================
FAIL: test_IcingDetected (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 375, in test_IcingDetected
    self.assertEqual(bytes(IcingDetected(example_value)), example_ls_packet)
AssertionError: b'"\x040x9b' != b'"\x01\x9b'

======================================================================
FAIL: test_LaserPRFCode (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 620, in test_LaserPRFCode
    self.assertEqual(bytes(LaserPRFCode(example_value)), example_ls_packet)
AssertionError: b'>\x82\xc6\xcf\x00\x00\x00\x00\x00\x00\x00\x0[203545 chars]\x00' != b'>\x02\xc6\xcf'

======================================================================
FAIL: test_SensorEastVelocity (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 763, in test_SensorEastVelocity
    self.assertEqual(bytes(SensorEastVelocity(example_value)), example_ls_packet)
AssertionError: b'P\x02\x80\x01' != b'P\x02\x00\x00'

======================================================================
FAIL: test_SensorEllipsoidHeightExtended (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 916, in test_SensorEllipsoidHeightExtended
    self.assertEqual(bytes(SensorEllipsoidHeightExtended(example_value)), example_ls_packet)
AssertionError: b'h\x0823456.24' != b'h\x03/\x92\x1e'

======================================================================
FAIL: test_SensorFieldOfViewName (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 629, in test_SensorFieldOfViewName
    self.assertEqual(bytes(SensorFieldOfViewName(example_value)), example_ls_packet)
AssertionError: b'?\x81\xd1\x00\x00\x00\x00\x00\x00\x00\x00\x0[797 chars]\x00' != b'?\x01\xd1'

======================================================================
FAIL: test_TargetErrorEstimateCE90 (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 474, in test_TargetErrorEstimateCE90
    self.assertEqual(bytes(TargetErrorEstimateCE90(example_value)), example_ls_packet)
AssertionError: b'-\t425.21515' != b'-\x02\x1a\x95'

======================================================================
FAIL: test_TargetErrorEstimateLE90 (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 483, in test_TargetErrorEstimateLE90
    self.assertEqual(bytes(TargetErrorEstimateLE90(example_value)), example_ls_packet)
AssertionError: b'.\t608.92309' != b'.\x02&\x11'

======================================================================
FAIL: test_TargetTrackGateHeight (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 465, in test_TargetTrackGateHeight
    self.assertEqual(bytes(TargetTrackGateHeight(example_value)), example_ls_packet)
AssertionError: b',\x1e\x00\x00\x00\x00\x00\x00\x00\x00\x00\x0[77 chars]\x00' != b',\x01\x0f'

======================================================================
FAIL: test_TargetWidthExtended (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 886, in test_TargetWidthExtended
    self.assertEqual(bytes(TargetWidthExtended(example_value)), example_ls_packet)
AssertionError: b'`\n13898.5463' != b'`\x03\x00\xd9*'

======================================================================
FAIL: test_UASLSVersionNumber (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 648, in test_UASLSVersionNumber
    self.assertEqual(bytes(UASLSVersionNumber(example_value)), example_ls_packet)
AssertionError: b'A\x00' != b'A\x01\xe8'

======================================================================
FAIL: test_WeaponFired (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 611, in test_WeaponFired
    self.assertEqual(bytes(WeaponFired(example_value)), example_ls_packet)
AssertionError: b'=\x81\xba\x00\x00\x00\x00\x00\x00\x00\x00\x0[705 chars]\x00' != b'=\x01\xba'

======================================================================
FAIL: test_WeaponLoad (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 602, in test_WeaponLoad
    self.assertEqual(bytes(WeaponLoad(example_value)), example_ls_packet)
AssertionError: b'<\x82\xaf\xd8\x00\x00\x00\x00\x00\x00\x00\x0[180029 chars]\x00' != b'<\x02\xaf\xd8'

======================================================================
FAIL: test_checksum (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 37, in test_checksum
    self.assertEqual(str(Checksum(value).value), interpretation)
AssertionError: '0x00000000000000000000000000000000000000000[87129 chars]0000' != '0xAA43'

----------------------------------------------------------------------
Ran 91 tests in 0.048s

FAILED (failures=16, errors=2)
```

**Comment by paretech** [comment]
@All4Gis I suggest starting at the top of the list of failing tests and working your way down. 

OutsideAirTemperature and TargetTrackGateWidth need abstract methods _domain and _range when subclassing from MappedElementParser.

GenericFlagData01, IcingDetected, LaserPRFCode, SensorFieldOfViewName, TargetErrorEstimateCE90, TargetErrorEstimateLE90, TargetTrackGateHeight,UASLSVersionNumber, WeaponFired, and WeaponLoad subclass from MappedElementParser.

Compare the parameters for SensorEastVelocity with those specified by MISB ST 0601 and resolve inconsistencies. 

Checksum backout changes to elementparser.BytesValue.

Ignore AlternatePlatformEllipsoidHeightExtended, DensityAltitudeExtended, SensorEllipsoidHeightExtended, and TargetWidthExtended. When this PR is complete, I expect these tests will be failing. There is additional development elsewhere in the codebase that needs to be completed before these will pass.

**Comment by All4Gis** [comment]
Commented AlternatePlatformEllipsoidHeightExtended, DensityAltitudeExtended, SensorEllipsoidHeightExtended, and TargetWidthExtended for pass test.

Pass all test except 

```
Finding files... done.
Importing test modules ... done.

======================================================================
ERROR: test_WeaponLoad (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 602, in test_WeaponLoad
    self.assertEqual(bytes(WeaponLoad(example_value)), example_ls_packet)
  File "D:\GitHub\klvdata\klvdata\element.py", line 64, in __bytes__
    return bytes(self.key) + bytes(self.length) + bytes(self.value)
  File "D:\GitHub\klvdata\klvdata\element.py", line 60, in length
    return ber_encode(len(self))
  File "D:\GitHub\klvdata\klvdata\element.py", line 68, in __len__
    return len(bytes(self.value))
  File "D:\GitHub\klvdata\klvdata\elementparser.py", line 157, in __bytes__
    return float_to_bytes(self.value, self._domain, self._range)
  File "D:\GitHub\klvdata\klvdata\common.py", line 142, in float_to_bytes
    dst_value = linear_map(value, src_domain=src_domain, dst_range=dst_range)
  File "D:\GitHub\klvdata\klvdata\common.py", line 117, in linear_map
    raise ValueError
ValueError

======================================================================
FAIL: test_OutsideAirTemperature (test.test_misb_0601.ParserSingleShort)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\GitHub\klvdata\test\test_misb_0601.py", line 420, in test_OutsideAirTemperature
    self.assertEqual(bytes(OutsideAirTemperature(example_value)), example_ls_packet)
AssertionError: b"'\x01\xd4" != b"'\x01T"

----------------------------------------------------------------------
Ran 87 tests in 0.058s

FAILED (failures=1, errors=1)
```

I don't have the correct range for WeaponLoad and OutsideAirTemperature,i think that convert 54 hex to Ascii and print the symbol T.

Check this when you can.
regards
